File: java/src/main/java/io/hron/ast/UntilSpec.java
package io.hron.ast;

/**
 * Represents an until date.
 *
 * @param kind the type of until specification
 * @param date the ISO date string (for ISO)
 * @param month the month (for NAMED)
 * @param day the day (for NAMED)
 */
public record UntilSpec(Kind kind, String date, MonthName month, int day) {

  /** The type of until specification. */
  public enum Kind {
    /** An ISO until date (e.g., 2026-12-31). */
    ISO,
    /** A named until date (e.g., dec 31). */
    NAMED
  }

  /**
   * Creates an ISO until specification.
   *
   * @param date the ISO date string (YYYY-MM-DD)
   * @return a new ISO until specification
   */
  public static UntilSpec iso(String date) {
    return new UntilSpec(Kind.ISO, date, null, 0);
  }

  /**
   * Creates a named until specification.
   *
   * @param month the month
   * @param day the day
   * @return a new named until specification
   */
  public static UntilSpec named(MonthName month, int day) {
    return new UntilSpec(Kind.NAMED, null, month, day);
  }
}

File: csharp/Hron/Ast/UntilSpec.cs
namespace Hron.Ast;

/// <summary>
/// Represents an until date.
/// </summary>
public sealed record UntilSpec(UntilSpecKind Kind, string? Date, MonthName? Month, int Day)
{
    /// <summary>
    /// Creates an ISO until specification.
    /// </summary>
    public static UntilSpec Iso(string date)
        => new(UntilSpecKind.Iso, date, null, 0);

    /// <summary>
    /// Creates a named until specification.
    /// </summary>
    public static UntilSpec Named(MonthName month, int day)
        => new(UntilSpecKind.Named, null, month, day);
}

/// <summary>
/// The type of until specification.
/// </summary>
public enum UntilSpecKind
{
    /// <summary>An ISO until date (e.g., 2026-12-31).</summary>
    Iso,
    /// <summary>A named until date (e.g., dec 31).</summary>
    Named
}
